pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -1308,36 +1308,44 @@ func (c *PodAffinityChecker) InterPodAffinityMatches(pod *v1.Pod, meta Predicate
 	return true, nil, nil
 }
 
-// podMatchesPodAffinityTerms checks if the "targetPod" matches the given "terms"
-// of the "pod" on the given "nodeInfo".Node(). It returns three values: 1) whether
-// targetPod matches all the terms and their topologies, 2) whether targetPod
-// matches all the terms label selector and namespaces (AKA term properties),
-// 3) any error.
-func (c *PodAffinityChecker) podMatchesPodAffinityTerms(pod, targetPod *v1.Pod, nodeInfo *schedulernodeinfo.NodeInfo, terms []v1.PodAffinityTerm) (bool, bool, error) {
-	if len(terms) == 0 {
-		return false, false, fmt.Errorf("terms array is empty")
-	}
-	props, err := getAffinityTermProperties(pod, terms)
+// podMatchesPodAffinityTerms checks if the "targetPod" matches the "checkAffinity (true:affinity,false:anti-affinity)"
+// of the "pod" on the given "nodeInfo".Node(). It returns four values: 1) whether
+// targetPod matches affnityTerms and their topologies based on the request of affinity or anti-affinity 2) whether targetPod
+// matches all the terms label selector and namespaces (AKA term properties), 3) any error.
+func (c *PodAffinityChecker) podMatchesPodAffinityTerms(pod, targetPod *v1.Pod, nodeInfo *schedulernodeinfo.NodeInfo, checkAffinity bool) (bool, bool, error) {
+	matchedTerms, affinityTermsNum, err := getMatchingAffinityTerms(pod, targetPod, checkAffinity)
 	if err != nil {
 		return false, false, err
 	}
-	if !podMatchesAllAffinityTermProperties(targetPod, props) {
-		return false, false, nil
+	termsSelectorMatched := len(matchedTerms) == affinityTermsNum
+	if checkAffinity && !termsSelectorMatched {
+		return false, false, err
 	}
-	// Namespace and selector of the terms have matched. Now we check topology of the terms.
 	targetPodNode, err := c.info.GetNodeInfo(targetPod.Spec.NodeName)
 	if err != nil {
-		return false, false, err
+		return false, termsSelectorMatched, err
 	}
-	for _, term := range terms {
+	if len(matchedTerms) == 0 {
+		return false, false, nil
+	}
```

Comment:
This can be merged to L1321 to be:

```go
// for PodAffinity, return false when not all terms are matched
// for PodAntiAffinity, return false when not any term is matched
if (checkAffinity && !termsSelectorMatched) || len(matchedTerms) == 0 {
	return false, termsSelectorMatched, err
}
```